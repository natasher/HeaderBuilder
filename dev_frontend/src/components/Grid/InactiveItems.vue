<template>
<section id = "mfn-hb-inactive-items">

  <div class = "header">
    <span class = "title">
      Inactive items
    </span>
  </div>

  <div class = "hb-items">
    <draggable
      class    = "hb-col"
      v-model  = "inactiveItems"
      key      = "inactiveItems"
      @change  = "onChange"
      :options = "{
        group: {
          name: 'items',
        },
        scroll: true,
        disabled: ( getGridStatus === 'off' || getGridStatus === 'auto')
                  ? true
                  : false,
      }" >

      <base-item
        v-for   = "( item, index ) in inactiveItems"
        row     = "inactiveItems"
        :item   = "item"
        :cellId = "index"
        :key    = "item.uuid" />

    </draggable>
  </div>

</section>
</template>

<script>
import { mapGetters, mapActions } from 'vuex'
import FontIcon  from '../BaseGeneric/FontIcon.vue'
import BaseItem  from '../BaseItem.vue'
import draggable from 'vuedraggable'

export default {

  name: 'InactiveItems',

  components: {
    FontIcon,
    BaseItem,
    draggable,
  },

  computed: {
    ...mapGetters( 'ui', [
      'getGridStatus',
    ]),

    ...mapGetters( 'inactiveItems', [
      'getInactiveItems',
    ]),

    inactiveItems: {
      get() {
        return this.getInactiveItems;
      },
      set() {}
    }
  },

  methods: {
    ...mapActions( 'inactiveItems', [
      'removeItem',
    ]),

    onChange( event ) {

      if ( event.removed ) {
        const cellId = event.removed.oldIndex

        this.removeItem({
          cellId : cellId,
        })
      }

    },
  },

}
</script>
