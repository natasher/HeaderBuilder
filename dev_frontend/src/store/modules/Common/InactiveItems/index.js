export default {

  namespaced: true,

  getters: {

    getInactiveItems( state, getters, rootState ) {
      const currentModule = rootState.general.currentStoreModule

      if ( currentModule !== '' ) {
        return rootState[ currentModule ].inactiveItems
      }
    },

  },

  actions: {

    /**
     * Insert the `item` at the end of `MobileModule/inactiveItems` Array
     * ( same at MobileStickyModule ).
     * Authors recommendation for Mobile devices is leave only 'logo' and 'menu' item for the header.
     * The User has possiblity to drag cloned items from `inactiveItems` to header grid back.
     *
     * @param {function} commit Synchronous invoke mutation.
     * @param {Object} rootState Access to global state.
     * @param {Object} payload Object that represents data sends to store
     *    - item {object}
     * @returns {Promise} Action resolves after mutation is commited.
     */
    putItem({ commit, rootState }, payload ) {
      return new Promise(( resolve ) => {
        commit( `${ rootState.general.currentStoreModule }/PUT_TO_INACTIVE`, payload, { root: true } )
        resolve()
      })
    },

    /**
     * Remove item object from inactiveItems array.
     *
     * @param {function} commit Synchronous invoke mutation.
     * @param {Object} rootState Access to global state.
     * @param {Object} payload Object that represents data sends to store
     *    - cellId {number} position where removed item lays.
     * @returns {Promise} Action resolves after mutation is commited.
     */
    removeItem({ commit, rootState }, payload ) {
      return new Promise(( resolve ) => {
        const currentStoreModule = rootState.general.currentStoreModule

        commit( `${ currentStoreModule }/REMOVE_ITEM_FROM_INACTIVE`, payload, { root: true })

        resolve()
      })
    },

    /**
     * Clear the `MobileModule{Sticky}/inactiveItems` cell by replacing its content with empty array.
     * Purpose: avoid duplicating items in the cells.
     * No payload is needed.
     *
     * @param {function} commit Synchronous invoke mutation.
     * @param {Object} rootState Access to global state.
     * @returns {Promise} Action resolves after mutation is commited.
     */
    clear({ commit, rootState }) {
      return new Promise(( resolve ) => {
        commit( `${ rootState.general.currentStoreModule }/CLEAR_INACTIVE_ITEMS`, void 0, { root: true } )
        resolve()
      })
    },

    /**
     * Copy inactive items from MobileModule to MobileStickyModule
     *
     * @param {function} dispatch Asynchronous invoke action.
     * @param {Object} rootState Access to global state.
     * @returns {Promise} Action resolves after mutation is commited.
     */
    copy({ dispatch, rootState }) {
      return new Promise(( resolve ) => {
        const inactiveItems = rootState.MobileModule.inactiveItems

        _.each( inactiveItems, function ( item ) {
          dispatch( 'putItem', { item: item })
        })

        resolve()
      })
    },

  },

}