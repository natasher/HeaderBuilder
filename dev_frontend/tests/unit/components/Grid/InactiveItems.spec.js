import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex          from 'vuex'
import merge         from 'lodash.merge'
import draggable     from 'vuedraggable'
import InactiveItems from '../../../../src/components/Grid/InactiveItems.vue'

const localVue = createLocalVue()
localVue.use( Vuex )

const testData = [
                {
                  "name": "social",
                  "uuid": "toWXiE1dIS",
                },
                {
                  "name": "menuIcon",
                  "uuid": "alBWLJBYfR"
                },
                {
                  "name": "button",
                  "uuid": "AJDV7xJMAN"
                }
              ]

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      ui: {
        namespaced: true,
        getters: {
          getGridStatus: jest.fn().mockReturnValue( 'custom' ),
        },
      },
      inactiveItems: {
        namespaced: true,
        getters: {
          getInactiveItems: jest.fn().mockReturnValue( [] )
        },
        actions: {
          removeItem: jest.fn(),
        }
      }
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    store: createStore(),
    stubs: {
      draggable,
    }
  }

  return shallowMount( InactiveItems, merge( defaultMountingOptions, overrides ) )
}

describe('InactiveItems.vue', () => {

  describe('snapshots:', () => {

    test('should match if no item is inactive', () => {
      const wrapper = createWrapper()

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match if 3 items is inactive', () => {
      const store = createStore({
        modules: {
          inactiveItems: {
            getters: {
              getInactiveItems: jest.fn().mockReturnValue( testData )
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })

  })

})
// the remove item test cant be done in unit test due to draggable library