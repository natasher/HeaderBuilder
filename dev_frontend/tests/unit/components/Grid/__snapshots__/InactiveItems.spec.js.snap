// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InactiveItems.vue snapshots: should match if 3 items is inactive 1`] = `
<section
  id="mfn-hb-inactive-items"
>
  <div
    class="header"
  >
    <span
      class="title"
    >
      
      Inactive items
    
    </span>
  </div>
   
  <div
    class="hb-items"
  >
    <div
      class="hb-col"
    >
      <baseitem-stub
        cellid="0"
        item="[object Object]"
        row="inactiveItems"
      />
      <baseitem-stub
        cellid="1"
        item="[object Object]"
        row="inactiveItems"
      />
      <baseitem-stub
        cellid="2"
        item="[object Object]"
        row="inactiveItems"
      />
    </div>
  </div>
</section>
`;

exports[`InactiveItems.vue snapshots: should match if no item is inactive 1`] = `
<section
  id="mfn-hb-inactive-items"
>
  <div
    class="header"
  >
    <span
      class="title"
    >
      
      Inactive items
    
    </span>
  </div>
   
  <div
    class="hb-items"
  >
    <div
      class="hb-col"
    />
  </div>
</section>
`;
